File: 1159-smallest-subsequence-of-distinct-characters/smallest-subsequence-of-distinct-characters.cpp
class Solution {
public:
    string smallestSubsequence(string s) {
        set<char>alpha;
       
        int size=alpha.size();
        int n=s.size();
        stack<char>st;
        map<char,int>mp;
        for(auto it:s){
            mp[it]++;
        }
        for(int i=0;i<s.size();i++){
            mp[s[i]]--;
            if(alpha.find(s[i])!=alpha.end()){
                continue;
            }
            while(!st.empty() && st.top()>s[i] && mp[st.top()]>0){
                alpha.erase(st.top());
                st.pop();
            }
            alpha.insert(s[i]); 
            st.push(s[i]);
                
            }
        string ans="";
        while(!st.empty()){
            ans+=st.top();
            st.pop();
        }
        reverse(ans.begin(),ans.end());
        
        return ans;
        
        
    }
};

File: 1159-smallest-subsequence-of-distinct-characters/smallest-subsequence-of-distinct-characters.java
class Solution {
    public String smallestSubsequence(String s) {
        int []freq=new int[26];
        for(int i=0;i<s.length();i++){
            freq[s.charAt(i)-'a']++;
        }
        Stack<Character>st=new Stack<>();
        boolean[] inStack = new boolean[26]; // for 'a' to 'z'
        for(int i=0;i<s.length();i++){
            if(inStack[s.charAt(i)-'a']==true){
                freq[s.charAt(i)-'a']--;
                continue;
            }
            while(!st.isEmpty() && st.peek()>=s.charAt(i)){
                if(freq[st.peek()-'a']==1)break;
                char ch=st.peek();
                freq[st.peek()-'a']--;
                inStack[ch-'a']=false;
                st.pop();

            }
            if(inStack[s.charAt(i)-'a']==false){
                st.push(s.charAt(i));
                inStack[s.charAt(i)-'a']=true;

            }
            
        }
        StringBuilder ans=new StringBuilder();
        

        while(!st.isEmpty())
        {
            ans.append(st.peek());
            st.pop();
        }
        String result=ans.reverse().toString();
        return result;

    }
}